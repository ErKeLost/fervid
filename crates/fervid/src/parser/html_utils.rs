use fervid_core::ElementKind;
use nom::{bytes::complete::{take_while1, take_while}, IResult};

// According to https://www.w3.org/TR/2011/WD-html5-20110525/syntax.html#elements-0
const VOID_TAGS: [&str; 16] = ["area", "base", "br", "col", "command", "embed", "hr", "img", "input", "keygen", "link", "meta", "param", "source", "track", "wbr"];
const RAW_TEXT_ELEMENTS: [&str; 2] = ["script", "style"];
const RCDATA_ELEMENTS: [&str; 2] = ["textarea", "title"];
const FOREIGN_ELEMENTS: [&str; 1] = ["svg"]; // todo

pub fn classify_element_kind(tag_name: &str) -> ElementKind {
  let tag_lowercase = &tag_name.to_lowercase();
  let tag_lowercase = tag_lowercase.as_str();
  if RCDATA_ELEMENTS.contains(&tag_lowercase) {
    ElementKind::RCData
  } else if FOREIGN_ELEMENTS.contains(&tag_lowercase) {
    ElementKind::Foreign
  } else if RAW_TEXT_ELEMENTS.contains(&tag_lowercase) {
    ElementKind::RawText
  } else if VOID_TAGS.contains(&tag_lowercase) {
    ElementKind::Void
  } else {
    ElementKind::Normal
  }
}

/**
 * U+0020 SPACE, U+0009 CHARACTER TABULATION (tab), U+000A LINE FEED (LF), U+000C FORM FEED (FF), and U+000D CARRIAGE RETURN (CR)
 * https://www.w3.org/TR/2011/WD-html5-20110525/common-microsyntaxes.html#space-character
 */
pub fn is_space_char(x: char) -> bool {
  x == ' ' || x == '\t' || x == '\n' || x == '\r' || x == '\u{000C}'
}

// todo allow more symbols as per W3 spec
pub fn is_valid_name_char(x: char) -> bool {
  // (x >= 'A' && x <= 'Z') || (x >= 'a' && x <= 'z')
  x.is_alphabetic()
}

pub fn html_name(input: &str) -> IResult<&str, &str> {
  // todo control dashes?? allow unicode??
  take_while1(|x: char| is_valid_name_char(x) || x == '-')(input)
}

pub fn space1(input: &str) -> IResult<&str, &str> {
  take_while1(is_space_char)(input)
}

pub fn space0(input: &str) -> IResult<&str, &str> {
  take_while(is_space_char)(input)
}
