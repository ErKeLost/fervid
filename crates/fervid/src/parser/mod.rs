pub mod attributes;
pub mod core;
pub mod html_utils;
pub mod sfc_blocks;
pub mod structs;
