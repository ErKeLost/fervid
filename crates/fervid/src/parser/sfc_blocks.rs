use fervid_core::{ElementNode, Node, HtmlAttribute, SfcBlock, SfcTemplateBlock, SfcStyleBlock, SfcScriptBlock};

/// Converts an untyped root block (which is always a [`Node::ElementNode`]) to an [`SfcBlock`],
/// which is a Vue SFC descriptor block
pub fn convert_node_to_typed<'e>(node: &'e ElementNode) -> SfcBlock<'e> {
    let starting_tag = &node.starting_tag;

    // Mutually exclusive flags
    let is_script = starting_tag.tag_name == "script";
    let is_template = !is_script && starting_tag.tag_name == "template";
    let is_style = !is_template && starting_tag.tag_name == "style";

    if !is_script && !is_template && !is_style {
        return SfcBlock::Custom(&node);
    }

    // Get `lang` attribute, which is common for all the Vue root blocks
    let lang = starting_tag.attributes.iter().find_map(|attr| match attr {
        HtmlAttribute::Regular {
            name: "lang",
            value,
        } => Some(*value),
        _ => None,
    });

    // First, check for template, this is already parsed
    if is_template {
        return SfcBlock::Template(SfcTemplateBlock {
            lang: lang.unwrap_or("html"),
            roots: &node.children,
        });
    }

    // For both `script` and `style`, the content is inside a TextNode
    let content = node
        .children
        .get(0)
        .and_then(|child| match child {
            Node::TextNode(text) => Some(*text),
            _ => None,
        })
        .unwrap_or("");

    if is_script {
        let is_setup = starting_tag
            .attributes
            .iter()
            .any(|attr| matches!(attr, HtmlAttribute::Regular { name: "setup", .. }));

        // TODO What should be done if for some reason `content` is an empty string?
        // This means that either parsing failed or the content is really empty
        // Maybe it should be checked in analyzer??

        return SfcBlock::Script(SfcScriptBlock {
            lang: lang.unwrap_or("js"),
            content,
            is_setup,
        });
    }

    if is_style {
        let is_scoped = starting_tag
            .attributes
            .iter()
            .any(|attr| matches!(attr, HtmlAttribute::Regular { name: "scoped", .. }));

        return SfcBlock::Style(SfcStyleBlock {
            lang: lang.unwrap_or("css"),
            content,
            is_scoped,
        });
    }

    unreachable!("All SFC block variants were handled")
}
