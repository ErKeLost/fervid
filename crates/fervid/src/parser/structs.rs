use super::{attributes::HtmlAttribute, html_utils::ElementKind};

#[derive(Debug, Clone)]
pub enum Node<'a> {
  ElementNode(ElementNode<'a>),

  TextNode(&'a str),
  DynamicExpression { value: &'a str, template_scope: u32 },
  CommentNode(&'a str)
}

#[derive(Debug, Clone)]
pub struct ElementNode<'a> {
  pub starting_tag: StartingTag<'a>,
  pub children: Vec<Node<'a>>,
  pub template_scope: u32
}

#[derive(Debug, Clone)]
pub struct StartingTag<'a> {
  pub tag_name: &'a str,
  pub attributes: Vec<HtmlAttribute<'a>>,
  pub is_self_closing: bool,
  pub kind: ElementKind
}
