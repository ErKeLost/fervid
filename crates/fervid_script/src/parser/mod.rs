mod swc;

pub use swc::{parse_javascript_module, parse_typescript_module};
