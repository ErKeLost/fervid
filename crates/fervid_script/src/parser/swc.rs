use swc_core::{
    common::{Span, SyntaxContext, BytePos},
    ecma::ast::{EsVersion, Module},
};
use swc_ecma_parser::{lexer::Lexer, EsConfig, Parser, StringInput, Syntax, TsConfig};

pub fn parse_javascript_module(
    input: &str,
    span_start: u32,
    es_config: EsConfig,
) -> Result<Module, swc_ecma_parser::error::Error> {
    let span = Span::new(
        BytePos(span_start),
        BytePos(span_start + input.len() as u32),
        SyntaxContext::empty(),
    );

    let lexer = Lexer::new(
        Syntax::Es(es_config),
        EsVersion::EsNext,
        StringInput::new(input, span.lo, span.hi),
        None,
    );

    let mut parser = Parser::new_from(lexer);

    parser.parse_module()
}

pub fn parse_typescript_module(
    input: &str,
    span_start: u32,
    ts_config: TsConfig,
) -> Result<Module, swc_ecma_parser::error::Error> {
    let span = Span::new(
        BytePos(span_start),
        BytePos(span_start + input.len() as u32),
        SyntaxContext::empty(),
    );

    let lexer = Lexer::new(
        Syntax::Typescript(ts_config),
        EsVersion::EsNext,
        StringInput::new(input, span.lo, span.hi),
        None,
    );

    let mut parser = Parser::new_from(lexer);

    parser.parse_typescript_module()
}
