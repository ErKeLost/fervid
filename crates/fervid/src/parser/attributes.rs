use fervid_core::{
    HtmlAttribute, VBindDirective, VCustomDirective, VDirective, VForDirective, VModelDirective,
    VOnDirective, VSlotDirective,
};
use nom::{
    branch::alt,
    bytes::complete::{tag, take_till},
    character::complete::char,
    combinator::fail,
    multi::many0,
    sequence::{delimited, preceded},
    IResult,
};

use crate::parser::html_utils::{html_name, space1};

fn parse_attr_value(input: &str) -> IResult<&str, &str> {
    delimited(char('"'), take_till(|c| c == '"'), char('"'))(input)
}

/// Parses a directive in form of `v-directive-name:directive-attribute.modifier1.modifier2`
///
/// Allows for shortcuts like `@` (same as `v-on`), `:` (`v-bind`) and `#` (`v-slot`)
fn parse_directive(input: &str) -> IResult<&str, HtmlAttribute> {
    let (input, prefix) = alt((tag("v-"), tag("@"), tag("#"), tag(":"), tag(".")))(input)?;

    // https://vuejs.org/api/built-in-directives.html#v-bind
    let mut is_bind_prop = false;
    let mut is_dynamic = false;

    // Determine directive name
    let mut has_argument = false;
    let (input, directive_name) = match prefix {
        "v-" => {
            let (input, name) = html_name(input)?;

            // next char is colon, shift input and set flag
            if let Some(':') = input.chars().next() {
                has_argument = true;
                (&input[1..], name)
            } else {
                (input, name)
            }
        }

        "@" => {
            has_argument = true;
            (input, "on")
        }

        ":" => {
            has_argument = true;
            (input, "bind")
        }

        "." => {
            has_argument = true;
            is_bind_prop = true;
            (input, "bind")
        }

        "#" => {
            has_argument = true;
            (input, "slot")
        }

        _ => {
            return Err(nom::Err::Error(nom::error::Error {
                code: nom::error::ErrorKind::Tag,
                input,
            }))
        }
    };

    // Read argument part if we spotted `:` earlier
    let (input, argument) = if has_argument {
        // Support v-slot:[slotname], v-bind:[attr], etc.
        let (input, arg) = if input.starts_with("[") {
            is_dynamic = true;

            delimited(char('['), html_name, char(']'))(input)?
        } else {
            html_name(input)?
        };

        (input, Some(arg))
    } else {
        (input, None)
    };

    #[cfg(dbg_print)]
    {
        println!();
        println!("Parsed directive {:?}", directive_name);
        println!("Has argument: {}, argument: {:?}", has_argument, argument);
    }

    // Read modifiers
    let (input, modifiers): (&str, Vec<&str>) =
        many0(preceded(char('.'), html_name))(input).unwrap_or((input, vec![]));

    // Value
    let (input, value) = if !input.starts_with('=') {
        (input, None)
    } else {
        let (input, value) = parse_attr_value(&input[1..])?;
        (input, Some(value))
    };

    macro_rules! fail {
        () => {
            // TODO: this fails at a very unexpected location,
            // but maybe it needs to rewind back to the start
            return fail(input);
        };
    }

    /// Unwrapping the value or failing
    macro_rules! expect_value {
        () => {
            if let Some(value) = value {
                value
            } else {
                fail!();
            }
        };
    }

    // Type the directive
    let directive = match directive_name {
        // Directives arranged by estimated usage frequency
        "bind" => {
            // Get flags
            let mut is_camel = false;
            let mut is_prop = is_bind_prop;
            let mut is_attr = false;
            for modifier in modifiers.iter() {
                match *modifier {
                    "camel" => is_camel = true,
                    "prop" => is_prop = true,
                    "attr" => is_attr = true,
                    _ => {}
                }
            }

            let value = expect_value!();

            VDirective::Bind(VBindDirective {
                argument,
                value,
                is_dynamic_attr: is_dynamic,
                is_camel,
                is_prop,
                is_attr,
            })
        }
        "on" => VDirective::On(VOnDirective {
            event: argument,
            handler: value,
            is_dynamic_event: is_dynamic,
            modifiers,
        }),
        "if" => {
            let value = expect_value!();
            VDirective::If(value)
        }
        "else-if" => {
            let value = expect_value!();
            VDirective::ElseIf(value)
        }
        "else" => VDirective::Else,
        "for" => {
            let value = expect_value!();

            let Some((iterator, iterable)) = split_itervar_and_iterable(value) else {
                fail!();
            };

            VDirective::For(VForDirective { iterable, iterator })
        }
        "model" => {
            let value = expect_value!();

            VDirective::Model(VModelDirective {
                argument,
                value,
                modifiers,
            })
        }
        "slot" => VDirective::Slot(VSlotDirective {
            slot_name: argument,
            value,
            is_dynamic_slot: is_dynamic
        }),
        "show" => {
            let value = expect_value!();
            VDirective::Show(value)
        }
        "html" => {
            let value = expect_value!();
            VDirective::Html(value)
        }
        "text" => {
            let value = expect_value!();
            VDirective::Text(value)
        }
        "once" => VDirective::Once,
        "pre" => VDirective::Pre,
        "memo" => {
            let value = expect_value!();
            VDirective::Memo(value)
        }
        "cloak" => VDirective::Cloak,

        // Custom
        _ => VDirective::Custom(VCustomDirective {
            name: directive_name,
            argument,
            modifiers,
            value,
        }),
    };

    Ok((input, HtmlAttribute::VDirective(directive)))
}

// fn parse_dynamic_attr(input: &str) -> IResult<&str, HtmlAttribute> {
//     let (input, directive) = parse_directive(input)?;

//     #[cfg(dbg_print)]
//     println!("Dynamic attr: directive = {:?}", directive);

//     /* Try taking a `=` char, early return if it's not there */
//     if !input.starts_with('=') {
//         return Ok((input, directive));
//     }

//     let (input, attr_value) = parse_attr_value(&input[1..])?;

//     #[cfg(dbg_print)]
//     println!("Dynamic attr: value = {:?}", attr_value);

//     match directive {
//         HtmlAttribute::VDirective(directive) => Ok((
//             input,
//             HtmlAttribute::VDirective(VDirective {
//                 value: Some(attr_value),
//                 ..directive
//             }),
//         )),

//         /* Not possible, because parse_directive returns a directive indeed */
//         _ => Err(nom::Err::Error(nom::error::Error {
//             code: nom::error::ErrorKind::Fail,
//             input,
//         })),
//     }
// }

fn parse_vanilla_attr(input: &str) -> IResult<&str, HtmlAttribute> {
    let (input, attr_name) = html_name(input)?;

    /* Support omitting a `=` char */
    let eq: Result<(&str, char), nom::Err<nom::error::Error<_>>> = char('=')(input);
    match eq {
        // consider omitted attribute as attribute name itself (as current Vue compiler does)
        Err(_) => Ok((
            input,
            HtmlAttribute::Regular {
                name: attr_name,
                value: &attr_name,
            },
        )),

        Ok((input, _)) => {
            let (input, attr_value) = parse_attr_value(input)?;

            #[cfg(dbg_print)]
            println!("Dynamic attr: value = {:?}", attr_value);

            Ok((
                input,
                HtmlAttribute::Regular {
                    name: attr_name,
                    value: attr_value,
                },
            ))
        }
    }
}

fn parse_attr(input: &str) -> IResult<&str, HtmlAttribute> {
    let (input, attr) = alt((parse_directive, parse_vanilla_attr))(input)?;

    #[cfg(dbg_print)]
    {
        println!("Attribute: {:?}", attr);
        println!("Remaining input: {:?}", input);
    }

    Ok((input, attr))
}

pub fn parse_attributes(input: &str) -> IResult<&str, Vec<HtmlAttribute>> {
    many0(preceded(space1, parse_attr))(input)
}

fn split_itervar_and_iterable<'a>(raw: &'a str) -> Option<(&'a str, &'a str)> {
    // Try guessing: `item in iterable`
    let mut split = raw.splitn(2, " in ");
    if let (Some(itervar), Some(iterable)) = (split.next(), split.next()) {
        return Some((itervar.trim(), iterable.trim()));
    }

    // Try `item of iterable`
    let mut split = raw.splitn(2, " of ");
    if let (Some(itervar), Some(iterable)) = (split.next(), split.next()) {
        return Some((itervar.trim(), iterable.trim()));
    }

    // Not valid
    None
}
