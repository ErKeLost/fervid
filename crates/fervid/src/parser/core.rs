extern crate nom;
use nom::branch::alt;
use nom::bytes::complete::{take_until1, take_until};
use nom::combinator::fail;
use nom::multi::many0;
use nom::sequence::{preceded, delimited};
use nom::{
  IResult,
  bytes::complete::tag,
  sequence::tuple
};
use std::str;

use crate::parser::html_utils::classify_element_kind;

use super::attributes::parse_attributes;
use super::html_utils::{html_name, space0};
use fervid_core::{ElementNode, StartingTag, Node, ElementKind, SfcBlock, SfcCustomBlock, HtmlAttribute, SfcTemplateBlock, SfcScriptBlock, SfcStyleBlock};

/// Parses the Vue Single-File Component
///
/// The Ok variant is a tuple, where:
/// - the `.0` element is the remaining input. It should be any trailing whitespace if parsing succeeded;
/// - the `.1` element is a vector of root blocks, i.e. all `<script>`, `<template>`, `<style>` and custom blocks.
///
/// This function does not modify whitespace inside the blocks.
///
/// To optimize template node, use [`crate::analyzer::ast_optimizer::optimize_ast`]
pub fn parse_sfc(input: &str) -> IResult<&str, Vec<SfcBlock>> {
  many0(parse_root_block)(input)
}

fn parse_root_block<'a>(input: &'a str) -> IResult<&'a str, SfcBlock<'a>> {
  // Remove leading space
  let input = input.trim_start();

  // Read starting tag
  let (input, starting_tag) = parse_element_starting_tag(input)?;

  // Shortcut
  let is_self_closing = starting_tag.is_self_closing;

  // Mutually exclusive flags
  let is_script = starting_tag.tag_name == "script";
  let is_template = !is_script && starting_tag.tag_name == "template";
  let is_style = !is_template && starting_tag.tag_name == "style";

  // Helper fn
  let read_raw_text = |input: &'a str| parse_text_node(input).map_or_else(
    move |_| (input, ""),
    move |v| {
      let Node::TextNode(content) = v.1 else {
        unreachable!("parse_text_node always returns a Node::TextNode")
      };
      (v.0, content)
    }
  );

  // Read custom blocks as raw text
  if !is_script && !is_template && !is_style {
    // Do not process anything if starting tag is self-closing
    if is_self_closing {
      return Ok((input, SfcBlock::Custom(SfcCustomBlock {
        starting_tag,
        content: ""
      })));
    }

    // Read raw text and end tag
    let (input, content) = read_raw_text(input);
    let (input, _end_tag) = parse_element_end_tag(input)?;

    return Ok((input, SfcBlock::Custom(SfcCustomBlock { starting_tag, content })));
  }

  // Get `lang` attribute, which is common for all the Vue root blocks
  let lang = starting_tag.attributes.iter().find_map(|attr| match attr {
    HtmlAttribute::Regular {
      name: "lang",
      value,
    } => Some(*value),
    _ => None,
  });

  // Parse children only for `<template>` root block
  if is_template {
    // Parser does not really look at it currently
    let lang = lang.unwrap_or("html");

    // Check for self-closing `<template />`. I see no reason why someone might do it, but still
    if is_self_closing {
      return Ok((input, SfcBlock::Template(SfcTemplateBlock {
        lang,
        roots: Vec::new()
      })));
    }

    // Parse children, this may Err (and is not handled yet)
    let (input, children) = parse_node_children(input)?;

    // End tag as well. No checks are present that it is the same
    // In the future, this checks may be implemented
    let (input, _end_tag) = parse_element_end_tag(input)?;

    return Ok((input, SfcBlock::Template(SfcTemplateBlock { lang, roots: children })));
  }

  // Read script
  if is_script {
    let lang = lang.unwrap_or("js");

    let is_setup = starting_tag
      .attributes
      .iter()
      .any(|attr| matches!(attr, HtmlAttribute::Regular { name: "setup", .. }));

    // Check self-closing
    if is_self_closing {
      return Ok((input, SfcBlock::Script(SfcScriptBlock { lang, content: "", is_setup })));
    }

    // Read raw text and end tag
    let (input, content) = read_raw_text(input);
    let (input, _end_tag) = parse_element_end_tag(input)?;

    return Ok((input, SfcBlock::Script(SfcScriptBlock { lang, content, is_setup })));
  }

  // Read style (basically same as script, but attributes are different)
  let lang = lang.unwrap_or("css");

  let is_scoped = starting_tag
    .attributes
    .iter()
    .any(|attr| matches!(attr, HtmlAttribute::Regular { name: "scoped", .. }));

  // Check self-closing
  if is_self_closing {
    return Ok((input, SfcBlock::Style(SfcStyleBlock { lang, content: "", is_scoped })));
  }

  // Read raw text and end tag (exactly like in script and custom blocks)
  let (input, content) = read_raw_text(input);
  let (input, _end_tag) = parse_element_end_tag(input)?;

  Ok((input, SfcBlock::Style(SfcStyleBlock { lang, content, is_scoped })))
}

fn parse_element_starting_tag(input: &str) -> IResult<&str, StartingTag> {
  let (input, (_, tag_name, attributes, _, ending_bracket)) = tuple((
    tag("<"),
    html_name,
    parse_attributes,
    space0,
    alt((tag(">"), tag("/>")))
  ))(input)?;

  #[cfg(dbg_print)]
  {
    println!("Tag name: {:?}", tag_name);
    println!("Attributes: {:?}", attributes);
  }

  Ok((input, StartingTag {
    tag_name,
    attributes,
    is_self_closing: ending_bracket == "/>",
    kind: classify_element_kind(&tag_name)
  }))
}

fn parse_element_end_tag(input: &str) -> IResult<&str, &str> {
  // eat any tag, because it may not match the start tag according to spec
  delimited(
    tag("</"),
    html_name,
    preceded(space0, tag(">"))
  )(input)
}

fn parse_dynamic_expression_node(input: &str) -> IResult<&str, Node> {
  let (input, expression_content) = parse_dynamic_expression(input)?;
  Ok((input, Node::DynamicExpression { value: expression_content.trim(), template_scope: 0 }))
}

// todo implement different processing ways:
// 1: parse node start and then recursively parse children
// 2: parse node start and seek the ending tag
pub fn parse_element_node(input: &str) -> IResult<&str, Node> {
  let (input, starting_tag) = parse_element_starting_tag(input)?;

  let early_return = matches!(starting_tag.kind, ElementKind::Void) || starting_tag.is_self_closing;

  if early_return {
    return Ok((
      input,
      Node::ElementNode(ElementNode {
        starting_tag,
        children: vec![],
        template_scope: 0
      })
    ));
  }

  let (input, children) = parse_node_children(input)?;

  // parse end tag
  let (input, end_tag) = parse_element_end_tag(input)?;

  // todo pass a stack of elements instead of a single tag
  // todo handle the error? soft/hard error -> either return Err or proceed and warn
  if end_tag != starting_tag.tag_name {
    println!("End tag does not match start tag: <{}> </{}>", &starting_tag.tag_name, &end_tag);
  }

  Ok((
    input,
    Node::ElementNode(ElementNode {
      starting_tag,
      children,
      template_scope: 0
    })
  ))
}

// parses {{ expression }}
fn parse_dynamic_expression(input: &str) -> IResult<&str, &str> {
  delimited(tag("{{"), take_until1("}}"), tag("}}"))(input)
}

fn parse_text_node(input: &str) -> IResult<&str, Node> {
  let mut iter = input.chars().peekable();
  let mut bytes_taken = 0;

  while let Some(ch) = iter.next() {
    if ch == '<' {
      break;
    };
    // todo support other delimiters
    if let ('{', Some('{')) = (ch, iter.peek()) {
      break;
    };
    bytes_taken += ch.len_utf8();
  }

  /* Return error if input length is too short */
  if bytes_taken == 0 {
    return fail(input);
  }

  let (text, input) = input.split_at(bytes_taken);

  Ok((
    input,
    Node::TextNode(text)
  ))
}

fn parse_comment_node(input: &str) -> IResult<&str, Node> {
  let (input, comment) = delimited(
    tag("<!--"),
    take_until("-->"),
    tag("-->")
  )(input)?;

  Ok((input, Node::CommentNode(comment)))
}

fn parse_node_children(input: &str) -> IResult<&str, Vec<Node>> {
  many0(alt((
    parse_dynamic_expression_node,
    parse_comment_node,
    parse_element_node,
    parse_text_node
  )))(input)
}
