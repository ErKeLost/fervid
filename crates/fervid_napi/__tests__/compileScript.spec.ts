import { describe, test, expect } from 'vitest'
import { parse as babelParse } from '@babel/parser'
import { Compiler, FervidCompileOptions } from '..'

describe('SFC analyze <script> bindings', () => {
// https://github.com/vuejs/core/blob/272ab9fbdcb1af0535108b9f888e80d612f9171d/packages/compiler-sfc/__tests__/compileScript.spec.ts#L1252-L1306
  describe('auto name inference', () => {
    test('basic', () => {
      const { content } = compile(
        `<script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
          id: ''
        },
      )
      expect(content).toMatch(`export default {
    __name: "FooBar"`)
      assertCode(content)
    })

    test('do not overwrite manual name (object)', () => {
      const { content } = compile(
        `<script>
          export default {
            name: 'Baz'
          }
          </script>
          <script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
          id: ''
        },
      )
      expect(content).not.toMatch(`name: 'FooBar'`)
      expect(content).toMatch(`name: 'Baz'`)
      assertCode(content)
    })

    test('do not overwrite manual name (call)', () => {
      const { content } = compile(
        `<script>
          import { defineComponent } from 'vue'
          export default defineComponent({
            name: 'Baz'
          })
          </script>
          <script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
          id: ''
        },
      )
      expect(content).not.toMatch(`name: 'FooBar'`)
      expect(content).toMatch(`name: 'Baz'`)
      assertCode(content)
    })
  })
})

// https://github.com/vuejs/core/blob/272ab9fbdcb1af0535108b9f888e80d612f9171d/packages/compiler-sfc/__tests__/utils.ts#L11-L24
function compile(src: string, options: FervidCompileOptions) {
  const compiler = new Compiler()
  const result = compiler.compileSync(src, options)

  if (result.errors.length) {
    console.warn(result.errors[0])
  }

  return {
    content: result.code
  }
}

function assertCode(code: string) {
  // parse the generated code to make sure it is valid
  try {
    babelParse(code, {
      sourceType: 'module',
      plugins: [
        'typescript',
        ['importAttributes', { deprecatedAssertSyntax: true }],
      ],
    })
  } catch (e: any) {
    console.log(code)
    throw e
  }
  expect(code).toMatchSnapshot()
}
