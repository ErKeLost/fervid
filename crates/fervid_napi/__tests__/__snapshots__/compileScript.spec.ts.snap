// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SFC analyze <script> bindings > auto name inference > basic 1`] = `
"import { toDisplayString as _toDisplayString } from "vue";
export default {
    __name: "FooBar",
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup () {
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC analyze <script> bindings > auto name inference > do not overwrite manual name (call) 1`] = `
"import { defineComponent } from 'vue';
import { toDisplayString as _toDisplayString } from "vue";
export default {
    name: 'Baz',
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup () {
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC analyze <script> bindings > auto name inference > do not overwrite manual name (object) 1`] = `
"import { toDisplayString as _toDisplayString } from "vue";
export default {
    name: 'Baz',
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup () {
        const a = 1;
        return {
            a
        };
    }
};
"
`;
